cmd/supernode/app/root.go

```diff
@@ -65,121 +122,180 @@ func setupFlags(cmd *cobra.Command, opt *Options) {
+	flagSet.Var(&defaultBaseProperties.MaxBandwidth, "max-bandwidth",
 		"network rate that supernode can use")
 
-	flagSet.IntVar(&opt.SchedulerCorePoolSize, "pool-size", opt.SchedulerCorePoolSize,
+	flagSet.Int("pool-size", defaultBaseProperties.SchedulerCorePoolSize,
 		"pool size is the core pool size of ScheduledExecutorService")
 
-	flagSet.BoolVar(&opt.EnableProfiler, "profiler", opt.EnableProfiler,
+	flagSet.Bool("profiler", defaultBaseProperties.EnableProfiler,
 		"profiler sets whether supernode HTTP server setups profiler")
 
-	flagSet.BoolVarP(&opt.Debug, "debug", "D", opt.Debug,
+	flagSet.BoolP("debug", "D", defaultBaseProperties.Debug,
 		"switch daemon log level to DEBUG mode")
 
-	flagSet.IntVar(&opt.PeerUpLimit, "up-limit", opt.PeerUpLimit,
+	flagSet.Int("up-limit", defaultBaseProperties.PeerUpLimit,
 		"upload limit for a peer to serve download tasks")
 
-	flagSet.IntVar(&opt.PeerDownLimit, "down-limit", opt.PeerDownLimit,
+	flagSet.Int("down-limit", defaultBaseProperties.PeerDownLimit,
 		"download limit for supernode to serve download tasks")
 
-	flagSet.StringVar(&opt.AdvertiseIP, "advertise-ip", "",
+	flagSet.String("advertise-ip", "",
 		"the supernode ip is the ip we advertise to other peers in the p2p-network")
 
-	flagSet.DurationVar(&opt.FailAccessInterval, "fail-access-interval", opt.FailAccessInterval,
+	flagSet.Duration("fail-access-interval", defaultBaseProperties.FailAccessInterval,
 		"fail access interval is the interval time after failed to access the URL")
 
-	flagSet.DurationVar(&opt.GCInitialDelay, "gc-initial-delay", opt.GCInitialDelay,
+	flagSet.Duration("gc-initial-delay", defaultBaseProperties.GCInitialDelay,
 		"gc initial delay is the delay time from the start to the first GC execution")
 
-	flagSet.DurationVar(&opt.GCMetaInterval, "gc-meta-interval", opt.GCMetaInterval,
+	flagSet.Duration("gc-meta-interval", defaultBaseProperties.GCMetaInterval,
 		"gc meta interval is the interval time to execute the GC meta")
 
-	flagSet.DurationVar(&opt.TaskExpireTime, "task-expire-time", opt.TaskExpireTime,
+	flagSet.Duration("task-expire-time", defaultBaseProperties.TaskExpireTime,
 		"task expire time is the time that a task is treated expired if the task is not accessed within the time")
 
-	flagSet.DurationVar(&opt.PeerGCDelay, "peer-gc-delay", opt.PeerGCDelay,
+	flagSet.Duration("peer-gc-delay", defaultBaseProperties.PeerGCDelay,
 		"peer gc delay is the delay time to execute the GC after the peer has reported the offline")
 
+	exitOnError(bindRootFlags(viper.GetViper()), "bind root command flags")
 }
 
-// runSuperNode prepares configs, setups essential details and runs supernode daemon.
-func runSuperNode() error {
-	// create home dir
-	if err := fileutils.CreateDirectory(options.HomeDir); err != nil {
-		return fmt.Errorf("failed to create home dir %s: %v", options.HomeDir, err)
+// bindRootFlags binds flags on rootCmd to the given viper instance.
+func bindRootFlags(v *viper.Viper) error {
```

Comment:
It would be better to eliminate all the redundant `if err := v.BindFlag...` using a for loop over something like this:

```go
[]struct{
    Key string
    Flag string
}{
    {"base.listenPort", "port"},
    // ...
}
```

Reply:
Good idea!